<a href="https://colab.research.google.com/github/datascience-hassan/Hassan_INFO4670_Spring2026/blob/main/Hassan_Abdule_Assignment_2_1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#1. Data Quality Check

In [93]:
!pip install numpy


1.1 Using Python (pandas, matplotlib, or seaborn), load and inspect the Assignment 2 dataset.

In [144]:
# Write your code here
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

df = pd.read_csv('Assignment2.csv')
df.head()


Column 1        Date  AveragePrice  Total Volume     4046       4225  \
0         0  12-27-2015          1.33      64236.62  1036.74   54454.85   
1         1  12-20-2015          1.35      54876.98   674.28   44638.81   
2         2  12-13-2015          0.93     118220.22   794.70  109149.67   
3         3   12-6-2015          1.08      78992.15  1132.00   71976.41   
4         4  11-29-2015          1.28      51039.60   941.48   43838.39   

     4770  Total Bags  Small Bags  Large Bags  XLarge Bags          type  \
0   48.16     8696.87     8603.62       93.25          0.0  conventional   
1   58.33     9505.56     9408.07       97.49          0.0  conventional   
2  130.50     8145.35     8042.21      103.14          0.0  conventional   
3   72.58     5811.16     5677.40      133.76          0.0  conventional   
4   75.78     6183.95     5986.26      197.69          0.0  conventional   

   year  region  
0  2015  Albany  
1  2015  Albany  
2  2015  Albany  
3  2015  Albany  
4  2015  Albany

Write code to explore the data distribution (e.g., region, type, year) and check whether there is any bias. Provide both the code and your interpretation.

In [145]:
# checking the descriptive statistics of numeric variables
df.describe()

Column 1  AveragePrice  Total Volume          4046          4225  \
count  18254.000000  18254.000000  1.825300e+04  1.825200e+04  1.825300e+04   
mean      24.228060      1.406073  8.504609e+05  2.929605e+05  2.950905e+05   
std       15.480994      0.402668  3.453189e+06  1.264891e+06  1.203996e+06   
min        0.000000      0.440000  8.456000e+01  0.000000e+00  0.000000e+00   
25%       10.000000      1.100000  1.084277e+04  8.544900e+02  3.006890e+03   
50%       24.000000      1.370000  1.073061e+05  8.640940e+03  2.905471e+04   
75%       38.000000      1.660000  4.329238e+05  1.109857e+05  1.500448e+05   
max       52.000000      3.250000  6.250565e+07  2.274362e+07  2.047057e+07   

               4770    Total Bags    Small Bags    Large Bags    XLarge Bags  \
count  1.825300e+04  1.825300e+04  1.825200e+04  1.825200e+04   18253.000000   
mean   2.283481e+04  2.395890e+05  1.821666e+05  5.432919e+04    3105.745759   
std    1.074528e+05  9.861401e+05  7.461204e+05  2.439469e+05   17691.015569   
min    0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00       0.000000   
25%    0.000000e+00  5.090410e+03  2.852128e+03  1.273700e+02       0.000000   
50%    1.849900e+02  3.973703e+04  2.634000e+04  2.646115e+03       0.000000   
75%    6.237960e+03  1.107744e+05  8.333329e+04  2.199632e+04     132.230000   
max    2.546439e+06  1.937313e+07  1.338459e+07  5.719097e+06  551693.650000   

               year  
count  18254.000000  
mean    2016.142161  
std        1.254196  
min     1904.000000  
25%     2015.000000  
50%     2016.000000  
75%     2017.000000  
max     2018.000000

In [146]:
region = (df["region"].value_counts(normalize=True) * 100).round(2)

print("region distribution (%)")
print(region)

print("number of regions:", len(region))





region distribution (%)
region
WestTexNewMexico       1.86
Albany                 1.85
BaltimoreWashington    1.85
Boise                  1.85
Boston                 1.85
Atlanta                1.85
California             1.85
Charlotte              1.85
Chicago                1.85
CincinnatiDayton       1.85
Columbus               1.85
DallasFtWorth          1.85
Denver                 1.85
Detroit                1.85
GrandRapids            1.85
GreatLakes             1.85
HarrisburgScranton     1.85
HartfordSpringfield    1.85
Houston                1.85
Indianapolis           1.85
Jacksonville           1.85
BuffaloRochester       1.85
LasVegas               1.85
LosAngeles             1.85
MiamiFtLauderdale      1.85
Louisville             1.85
Nashville              1.85
NewOrleansMobile       1.85
NewYork                1.85
Midsouth               1.85
NorthernNewEngland     1.85
Orlando                1.85
Philadelphia           1.85
PhoenixTucson          1.85
Pittsburgh       

The dataset includes 54 regions, and each region represents about ~1.85% - 1.86% of the data. This shows the data is evenly distributed across regions

In [147]:
type_dist = (df["type"].value_counts(normalize=True, dropna=False) * 100).round(2)

print("type distribution (%)")
print(type_dist)

print("number of types:", len(type_dist))


type distribution (%)
type
organic         50.00
conventional    49.99
NaN              0.01
Name: proportion, dtype: float64
number of types: 3


Type is balanced: organic = 50.00% and conventional = 49.99%, so there is no type bias.

In [148]:
year_counts = df["year"].value_counts(dropna=False).sort_index()

print("\nyear counts")
print(year_counts)




year counts
year
1904       1
2015    5615
2016    5616
2017    5722
2018    1300
Name: count, dtype: int64


The dataset is mostly from 2015 - 2018 about 5600–5722 records each.

2018 has much fewer records 1300) so the data is imbalanced by year

And there is one record that has 1300 and i think that was a mistake

1.2 Write Python code to check for duplicate rows and missing values in the dataset. Show the number of duplicates and missing values for each column. Then, explain (in comments or markdown) how you would handle these issues (e.g., drop, impute, or replace).

In [149]:
# Write your code here

missing_rate = df.isnull().sum().sum()
print('TOTAL missing values, ', missing_rate)
print('Duplicates' , df.duplicated().sum())


TOTAL missing values,  12
Duplicates 2


we usully ingore very small nun value like 1 percent of the data since missing is tiny i would drop the few missing rows, but if i was modeling then i would fill with median and the type with mode to keep the data either option is fine

1.3 Use Python code to print the number of rows and columns in the dataset (e.g., with df.shape). Based on the dataset size, explain (briefly) whether you think the dataset is sufficient for training a machine learning model.

In [150]:
# Write your code here
df.shape
#There are 18254 rows and 14 columns

(18254, 14)

With 18,254 rows and 14 columns, this is generally enough data to train a machine learning model, especially for common models like linear/logistic regression, decision trees, random forest, and gradient boosting.is. decent size

#2. Data Cleaning and Preprocessing

2.1 Remove the first column or “Column 1” from the dataset. Treat the ‘year’ variable as nominal.

In [151]:
# write your code here
df = df.drop(columns=["Column 1"])

# Treat the 'year' variable as nominal
df['year'] = df['year'].astype('object')
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18254 entries, 0 to 18253
Data columns (total 13 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Date          18254 non-null  object 
 1   AveragePrice  18254 non-null  float64
 2   Total Volume  18253 non-null  float64
 3   4046          18252 non-null  float64
 4   4225          18253 non-null  float64
 5   4770          18253 non-null  float64
 6   Total Bags    18253 non-null  float64
 7   Small Bags    18252 non-null  float64
 8   Large Bags    18252 non-null  float64
 9   XLarge Bags   18253 non-null  float64
 10  type          18253 non-null  object 
 11  year          18254 non-null  object 
 12  region        18254 non-null  object 
dtypes: float64(9), object(4)
memory usage: 1.8+ MB


2.2 Check for duplicate values and remove them.

In [152]:
# Write your code here
df = df.drop_duplicates()

#check for duplicates
print("how many duplicates left", df.duplicated().sum())


how many duplicates left 0


2.3 Check for missing values. If a data record (row) only has a few missing values, replace the missing values with the median of the column feature in that specific “Region” variable. If most column values in a data record are missing, remove the data record.

In [153]:
df = df.fillna(df.median(numeric_only=True))

df["type"] = df["type"].fillna(df["type"].mode()[0])


missing_values = df.isnull().sum().sum()
print('TOTAL missing values:', missing_values)





TOTAL missing values: 0


2.4 Find the correlation between the variables and describe how the correlated values among the variables impact the model accuracy.


In [154]:
# Write your code here
corr = df.corr(numeric_only=True)
print(corr)

# The correlation matrix shows several very high correlations
# sometimes linear models like linear/logistic regression less stable and sometimes reduce accuracy.

              AveragePrice  Total Volume      4046      4225      4770  \
AveragePrice      1.000000     -0.192772 -0.208334 -0.172948 -0.179463   
Total Volume     -0.192772      1.000000  0.977863  0.974181  0.872203   
4046             -0.208334      0.977863  1.000000  0.926111  0.833390   
4225             -0.172948      0.974181  0.926111  1.000000  0.887856   
4770             -0.179463      0.872203  0.833390  0.887856  1.000000   
Total Bags       -0.177107      0.963047  0.920057  0.905788  0.792316   
Small Bags       -0.174748      0.967238  0.925280  0.916032  0.802734   
Large Bags       -0.172958      0.880641  0.838646  0.810017  0.698473   
XLarge Bags      -0.117608      0.747159  0.699379  0.688811  0.679863   

              Total Bags  Small Bags  Large Bags  XLarge Bags  
AveragePrice   -0.177107   -0.174748   -0.172958    -0.117608  
Total Volume    0.963047    0.967238    0.880641     0.747159  
4046            0.920057    0.925280    0.838646     0.699379  
422

#3. Exploratory Data Analysis (EDA)


3.1 Describe the variables
- Describe all variables in the dataset.
- For continuous variables: report **range (min, max), mean, median, and distribution**.
- For categorical variables: list unique values.

In [155]:
# Write your code here
df.describe()

num_cols = df.select_dtypes(include="number").columns

for col in num_cols:
    mn = df[col].min()
    mx = df[col].max()
    mean = df[col].mean()
    med = df[col].median()

    if mean > med:
        dist = "Right-skewed (few very large values)"
    elif mean < med:
        dist = "Left-skewed (few very small values)"
    else:
        dist = "Roughly symmetric"

    print(f"\n{col}")
    print(f"Range (min, max): ({mn:.2f}, {mx:.2f})")
    print(f"Mean: {mean:.2f}")
    print(f"Median: {med:.2f}")
    print(f"Distribution: {dist}")








AveragePrice
Range (min, max): (0.44, 3.25)
Mean: 1.41
Median: 1.37
Distribution: Right-skewed (few very large values)

Total Volume
Range (min, max): (84.56, 62505646.52)
Mean: 850511.59
Median: 107354.25
Distribution: Right-skewed (few very large values)

4046
Range (min, max): (0.00, 22743616.17)
Mean: 292961.28
Median: 8643.20
Distribution: Right-skewed (few very large values)

4225
Range (min, max): (0.00, 20470572.61)
Mean: 295107.97
Median: 29056.73
Distribution: Right-skewed (few very large values)

4770
Range (min, max): (0.00, 2546439.11)
Mean: 22836.03
Median: 184.99
Distribution: Right-skewed (few very large values)

Total Bags
Range (min, max): (0.00, 19373134.37)
Mean: 239603.01
Median: 39738.53
Distribution: Right-skewed (few very large values)

Small Bags
Range (min, max): (0.00, 13384586.80)
Mean: 182168.14
Median: 26351.61
Distribution: Right-skewed (few very large values)

Large Bags
Range (min, max): (0.00, 5719096.61)
Mean: 54329.45
Median: 2647.27
Distribution: R

3.2 Inspect the earliest recorded date
- Find the earliest `Date`.
- Check if there are avocado prices recorded from the earliest date up to 2010.
- Comment: does the earliest data point look reasonable? Keep or remove?

In [156]:
# Write your code here
df["Date"] = pd.to_datetime(df["Date"],errors="coerce")
earliest_date = df["Date"].min()
print("Earliest Date:", earliest_date)

pre2011 = df[df["Date"] <= pd.Timestamp("2010")]
print("Records with Date <= 2010:", len(pre2011))
print(pre2011[["Date","year","region","AveragePrice"]].head())

# 1904 is not reasonable shoud be removed



Earliest Date: 1904-01-21 00:00:00
Records with Date <= 2010: 1
            Date  year            region  AveragePrice
18250 1904-01-21  1904  WestTexNewMexico          1.87


3.3 Highest average price
- Find the highest value in "AveragePrice".
- Report which region it belongs to.
- Describe how you obtained the result.

In [157]:
# Write your code here
df[df["AveragePrice"] == df["AveragePrice"].max()] # thats the highest value in AveragePrice

# it belongs to sanFrancisco region
# I used max() to find the highest value in AveragePrice, then filtered the dataset to the row where AveragePrice equals that maximum and read the region value







Date  AveragePrice  Total Volume     4046      4225  4770  \
14125 2016-10-30          3.25      16700.94  2325.93  11142.85   0.0   

       Total Bags  Small Bags  Large Bags  XLarge Bags     type  year  \
14125     3232.16     3232.16         0.0          0.0  organic  2016   

             region  
14125  SanFrancisco

3.4 Highest total volume
- Find the highest total volume of avocados.
- Report which region it belongs to.
- Describe how you obtained the result.

In [158]:
# Write your code here
df[df["Total Volume"] == df["Total Volume"].max()]

# it belongs to TotalUS
# I used .max() to get the highest value in Total Volume, then filtered the dataframe to the row where Total Volume equals that max and read the region

Date  AveragePrice  Total Volume        4046         4225  \
9097 2018-02-04          0.87   62505646.52  21620180.9  20445501.03   

            4770   Total Bags  Small Bags  Large Bags  XLarge Bags  \
9097  1066830.22  19373134.37  13384586.8  5719096.61    269450.96   

              type  year   region  
9097  conventional  2018  TotalUS